### PR TITLE
Reworked descriptor (setup + hash) system, added new GaussianDescriptorSet code

### PR DESCRIPTION
**Pull request by EricMusa**
The GaussianDescriptorSet (GDS) encompasses everything that the Gaussian descriptor parameters dict (commonly defined as `Gs` in example/test scripts) did and how it was processed. The config setting `config["dataset"]["fp_params"]` can be set to either the Gs dict or a GDS - the resulting Amptorch model will function properly both ways. 

The purpose of the GDS is to allow greater flexibility when defining fingerprints. The Gs dict allows descriptors to be defined only combinatorically, not individually. The GDS allows for defining descriptors via either method. 

**Work is coming soon for automating the creation of ACSFdescriptors.**

### CODE REVIEW

**Review by EricMusa — COMMENTED** [review]
Mostly implementation changes. Added descriptor/Gaussian/descriptor_set.py, tests/gaussian_descriptor_set_test.py, and examples/custom_descriptor_example. Also, a few extraneous changes to the order in which packages are imported were made? probably from pre-commit hooks.

File: amptorch/trainer.py [hunk 1]
```diff
@@ -221,6 +229,7 @@ def train(self, raw_data=None):
         self.net.fit(self.train_dataset, None)
         elapsed_time = time.time() - stime
         print(f"Training completed in {elapsed_time}s")
+        return elapsed_time
```
Comment (mshuaibii) [inline amptorch/trainer.py:232]: We can probably remove this, I imagine you had it for testing purposes.

File: amptorch/descriptor/Gaussian/__init__.py [hunk 2]
```diff
@@ -32,25 +36,32 @@ def __init__(self, Gs, elements, cutoff_func="cosine", gamma=None):
...
+                        element
+                    ] = self._prepare_descriptor_parameters_element(
+                        self.Gs["default"], self.element_indices
+                    )
+                else:
+                    raise NotImplementedError(
+                        "Symmetry function parameters not defined properly"
+                    )
+        elif hasattr(self.Gs, "descriptor_setup"):
+            self.descriptor_setup = self.Gs.descriptor_setup
```
Comment (mshuaibii) [inline amptorch/descriptor/Gaussian/__init__.py:59]: To verify, in the custom case `self.Gs` is a `GaussianDescriptorSet` object, correct?

File: examples/custom_descriptor_example.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+g4_zetas = [1.0, 4.0, 1.0, 4.0]
+g4_gammas = [1.0, 1.0, -1.0, -1.0]
+gds_custom.batch_add_descriptors(
+    4, g4_etas, g4_zetas, g4_gammas
+)  # this opens opportunities for finely-tuned, custom descriptors
+
+print("both GaussianDescriptorSets are equivalent!")
+print("combo hash", gds_combo.descriptor_setup_hash)
+print("custom hash", gds_custom.descriptor_setup_hash)
+print("gds_combo == gds_custom:", gds_combo == gds_custom)
```
Comment (mshuaibii) [inline examples/custom_descriptor_example.py:62]: Instead, can you make this example demonstrate what the custom `GDS` allows you to do rather than comparing it to the old (you already do that in the test script)? For instance construct fingerprints other than the default method.

**Review by mshuaibii — CHANGES_REQUESTED** [review]
Looks good - some minor comments. Also, your commit history seems a little outdated with master, can you rebase this. Thanks!

Should I expect the automated fingerprint generation as part of this PR or that'll come in a different PR.

File: amptorch/trainer.py (hunk 1 above)
Comment (EricMusa) [reply amptorch/trainer.py:232]: done

File: amptorch/descriptor/Gaussian/__init__.py (hunk 2 above)
Comment (EricMusa) [reply amptorch/descriptor/Gaussian/__init__.py:59]: yes, I left it as this `hasattr` instead of `isinstance` in case I came up with a different, additional system like the `GDS` so the same attribute could be used

File: examples/custom_descriptor_example.py (hunk 3 above)
Comment (EricMusa) [reply examples/custom_descriptor_example.py:62]: yessir, good idea

**Comment by EricMusa** [comment]
Fixed the `return elapsed_time` thing and revised the GDS example script to show custom descriptor capability. Automated features will be coming soon, I've not implemented automation for G4/G5 yet, only G2.

I don't think I've been doing VCS correctly, when I go to rebase my master with your upstream, I have 97 commits to go through and resolve conflicts with. I'll get around to that when I have time in the next few days

What should I be doing to avoid this? I can't imagine it's supposed to be as hard as it seems... Should I be rebasing my forked branches with your upstream often?

**Comment by mshuaibii** [comment]
> Fixed the `return elapsed_time` thing and revised the GDS example script to show custom descriptor capability. Automated features will be coming soon, I've not implemented automation for G4/G5 yet, only G2.
> 
> I don't think I've been doing VCS correctly, when I go to rebase my master with your upstream, I have 97 commits to go through and resolve conflicts with. I'll get around to that when I have time in the next few days
> 
> What should I be doing to avoid this? I can't imagine it's supposed to be as hard as it seems... Should I be rebasing my forked branches with your upstream often?

Thanks! No I don't think you need to be rebasing so often. It looks like you worked on this branch without merging from upstream first. Your workflow should always begin by merging with upstream, make a new branch, work on your feature, create a PR, delete branch(optional), repeat. Not sure the cleanest way to resolve this existing PR, you can try doing the above steps and moving the changes of this PR into that branch and PR that branch instead.

**Comment by EricMusa** [comment]
Alright, I might have to refork, I have no idea what's going on. I spent an hour rebasing my master branch (100 commits worth of conflicts to resolve) only to have half of my edits overwritten by old commits. Then, once I recovered all my code and made sure the tests pass, I checked to see how my master is w.r.t. upstream/master and if I'd have to rebase again, and now I'm 100 commits behind again. I think it was because I tried merging with a pre-rebase branch to recover my old code...

Honestly I'm at a loss. Might be easier just to copy all my code, refork, branch, copy+paste, then re-PR. Will put a hell of a lot fewer commit messages on the history too.

If you have any advice/concerns please LMK!

**Comment by mshuaibii** [comment]
> Alright, I might have to refork, I have no idea what's going on. I spent an hour rebasing my master branch (100 commits worth of conflicts to resolve) only to have half of my edits overwritten by old commits. Then, once I recovered all my code and made sure the tests pass, I checked to see how my master is w.r.t. upstream/master and if I'd have to rebase again, and now I'm 100 commits behind again. I think it was because I tried merging with a pre-rebase branch to recover my old code...
> 
> Honestly I'm at a loss. Might be easier just to copy all my code, refork, branch, copy+paste, then re-PR. Will put a hell of a lot fewer commit messages on the history too.
> 
> If you have any advice/concerns please LMK!

Hmm yeah that sounds annoying, sorry about that. I had cleaned up the commit history a while back, depending when you forked your branch the issues you may be getting could be coming from that. I think reforking, branching, copy/pasting, the PR would clean this up for sure and ensure your history is consistent with master.

**Comment by EricMusa** [comment]
Abandon ship :(
